auth/api/iam/s2s_vptoken.go

```diff
@@ -19,49 +19,180 @@
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: invalid JSON",
+			InternalError: err,
+		}
+	}
 
-func (s serviceToService) validateVPToken(params map[string]string) (string, error) {
-	submission := params["presentation_submission"]
-	scope := params["scope"]
-	vp_token := params["vp_token"]
-	if submission == "" || scope == "" || vp_token == "" {
-		// TODO: right error response
-		return "", errors.New("missing required parameters")
-	}
-	// TODO: https://github.com/nuts-foundation/nuts-node/issues/2418
-	// TODO: verify parameters
-	return scope, nil
+	// Check VP signer == VC subject (presenter is holder of VCs)
+	if vp.Holder == nil {
+		// Is this even possible?
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "verifiable presentation is invalid: no holder",
+		}
+	}
+	err = credential.VerifyPresenterIsHolder(vp)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "verifiable presentation is invalid: " + err.Error(),
+		}
+	}
+
+	// Validate the presentation submission:
+	// 1. Resolve presentation definition for the requested scope
+	// 2. Take VCs mapped by the presentation submission
+	// 3. Match the VCs from (2) to the definition from (1). This should yield the same list of credentials.
+	//    This actually creates a new submission.
+	presentationDefinition := s.auth.PresentationDefinitions().ByScope(scope)
+	if presentationDefinition == nil {
+		return nil, OAuth2Error{
+			Code:        InvalidScope,
+			Description: fmt.Sprintf("unsupported scope for presentation exchange: %s", scope),
+		}
+	}
+	submissionCredentials, err := submission.Credentials(vp)
+	if err != nil {
+		return nil, err
+	}
+	// Match the credentials from the presentation submission with the presentation definition we got through the scope
+	_, matchedCredentials, err := presentationDefinition.Match(submissionCredentials)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: re-evaluation of the credentials against the presentation definition fails",
+			InternalError: err,
+		}
+	}
+
+	// Check signatures and VC issuer trust
+	issueTime := time.Now()
+	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "verifiable presentation is invalid",
+			InternalError: err,
+		}
+	}
+
+	submissionCredentialsJSON, _ := json.Marshal(submissionCredentials)
+	matchedCredentialsJSON, _ := json.Marshal(matchedCredentials)
+	if !bytes.Equal(submissionCredentialsJSON, matchedCredentialsJSON) {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "presentation_submission parameter is invalid: re-evaluation of the credentials against the presentation definition yields different credentials (bug or attempted hack)",
+		}
+	}
```

Comment:
I think this test can fail for several reasons even if `submissionCredentialsJSON` and `matchedCredentialsJSON` contain the same set (different order of VCs, different order of the elements in the JSON). Can we assume that the arrays are the same since `presentationDefinition.Match` did not return an error?

Reply:
Discussed: we need to introduce a reusable submission validation function